node/libs/crypto/src/secp256k1/mod.rs

```diff
@@ -0,0 +1,278 @@
+        Ok(PublicKey(vk))
+    }
+
+    /// Ensure the recovered public key is the expected one.
+    fn verify_pk(expected: &PublicKey, recovered: &PublicKey) -> anyhow::Result<()> {
+        anyhow::ensure!(
+            expected == recovered,
+            "PublicKey mismatch: expected {}, got {}",
+            hex::encode(expected.encode()),
+            hex::encode(recovered.encode())
+        );
+        Ok(())
+    }
+}
+
+impl ByteFmt for Signature {
+    fn decode(bytes: &[u8]) -> anyhow::Result<Self> {
+        anyhow::ensure!(
+            bytes.len() == SIGNATURE_LENGTH,
+            "unexpected signature length: {}",
+            bytes.len()
+        );
+        let recid = normalize_recovery_id(bytes[64]);
+        let Some(recid) = k256::ecdsa::RecoveryId::from_byte(recid) else {
+            bail!("unexpected recovery ID: {}", bytes[64]);
+        };
+        let sig = k256::ecdsa::Signature::from_slice(&bytes[..64])?;
+        Ok(Self { sig, recid })
+    }
+
+    fn encode(&self) -> Vec<u8> {
+        let mut bz = vec![0u8; SIGNATURE_LENGTH];
+        let (r, s) = self.sig.split_bytes();
+        bz[..32].copy_from_slice(&r);
+        bz[32..64].copy_from_slice(&s);
+        bz[64] = self.recid.to_byte() + 27;
+        bz
+    }
+}
+
+impl Hash for Signature {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        state.write(&self.encode())
+    }
+}
+
+impl PartialOrd for Signature {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Signature {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        ByteFmt::encode(self).cmp(&ByteFmt::encode(other))
+    }
+}
+
+/// Seclp256k1 multi-sig
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Default)]
+pub struct AggregateSignature(Vec<Signature>);
+
+impl AggregateSignature {
+    /// Add a signature to the aggregate.
+    pub fn add(&mut self, sig: Signature) {
+        self.0.push(sig)
+    }
+
+    /// Verifies an aggregated signature for multiple messages against the provided list of public keys.
+    ///
+    /// Verification fails if the total number of messages and signatures do not match.
+    ///
+    /// This method was added to mimic the bn254 version, which used BLS signature aggregation and did
+    /// not rely on message and key ordering; it supported different messages from each signatory as well.
+    ///
+    /// Here we have a simple list of signatures, and we cannot rely on them having been added in the same
+    /// order in which they are going to be verified, therefore we have to try every message against every
+    /// signature.
+    ///
+    /// The method assumes that there are no repeated pairs in the input, ie. that every signature is used exactly once.
+    pub fn verify_hash<'a>(
+        &self,
+        hashes_and_pks: impl Iterator<Item = (&'a [u8], &'a PublicKey)>,
```

Comment:
Is this function directly called from outside the crate? If it is, consider exposing a non-generic version of this function instead - we have compilation optimizations enabled for the zksync_consensus_crypto crate, even in debug builds - if this function gets inlined within other crate the optimizations won't get applied.

Reply:
It is called, yes. It would be awkward if it's a public type and it couldn't be called. I made it non-generic the same way as bn254 is.